Release 6.49.1 breaks @example in user code

In PINT we have a number of `hypothesis` tests, some of which used to fail and thus have `@example` attached. These were constructed in the usual way, by pasting them in from `hypothesis` output. This was a while ago! But with release 6.49.1, presumably the part about fixing inconsistencies in argument handling between `given` and `example`. 

The release that seems to have caused the problem: https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-6.49.0...hypothesis-python-6.49.1

The problems we are seeing: https://github.com/nanograv/PINT/runs/7203801688?check_suite_focus=true

Specifically:
```
________________________ test_d2tf_tf2d_roundtrip[9-1] _________________________
ndp = 9, k = 1
    @given(floats(-2, 2, allow_nan=False))
>   @example(ndp=10, f=1.3322676295501882e-15)
/home/runner/work/PINT/PINT/tests/test_precision.py:732: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
state = <hypothesis.core.StateForActualGivenExecution object at 0x7ff9bc348e80>
wrapped_test = <function accept.<locals>.test_d2tf_tf2d_roundtrip at 0x7ff9d15adbd0>
arguments = (), kwargs = {'k': 1, 'ndp': 9}
original_sig = <Signature (ndp, k, f)>
    def execute_explicit_examples(state, wrapped_test, arguments, kwargs, original_sig):
        posargs = [
            p.name
            for p in original_sig.parameters.values()
            if p.kind is p.POSITIONAL_OR_KEYWORD
        ]
        for example in reversed(getattr(wrapped_test, "hypothesis_explicit_examples", ())):
            if example.args:
                assert not example.kwargs
                if len(example.args) > len(posargs):
                    raise InvalidArgument(
                        "example has too many arguments for test. Expected at most "
                        f"{len(posargs)} but got {len(example.args)}"
                    )
                example_kwargs = dict(zip(posargs[::-1], example.args[::-1]))
            else:
                example_kwargs = dict(example.kwargs)
>           bound = original_sig.bind(*arguments, **example_kwargs, **kwargs)
E           TypeError: inspect.Signature.bind() got multiple values for keyword argument 'ndp'
/home/runner/work/PINT/PINT/.tox/py310-test-cov/lib/python3.10/site-packages/hypothesis/core.py:373: TypeError
```

If our `@example` syntax is somehow obsolete, we can try to fix it. Does `hypothesis` now suggest examples with syntax that works?

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,4 @@
+RELEASE_TYPE: patch
+
+This patch improves the error messages in :func:`@example() <hypothesis.example>`
+argument validation following the recent release of :ref:`6.49.1 <v6.49.1>`.

--- a/hypothesis-python/src/hypothesis/core.py
+++ b/hypothesis-python/src/hypothesis/core.py
@@ -359,29 +359,50 @@ def execute_explicit_examples(state, wrapped_test, arguments, kwargs, original_s
     ]
 
     for example in reversed(getattr(wrapped_test, "hypothesis_explicit_examples", ())):
+        # All of this validation is to check that @example() got "the same" arguments
+        # as @given, i.e. corresponding to the same parameters, even though they might
+        # be any mixture of positional and keyword arguments.
         if example.args:
             assert not example.kwargs
+            if any(
+                p.kind is p.POSITIONAL_ONLY for p in original_sig.parameters.values()
+            ):
+                raise InvalidArgument(
+                    "Cannot pass positional arguments to @example() when decorating "
+                    "a test function which has positional-only parameters."
+                )
             if len(example.args) > len(posargs):
                 raise InvalidArgument(
                     "example has too many arguments for test. Expected at most "
                     f"{len(posargs)} but got {len(example.args)}"
                 )
-            example_kwargs = dict(zip(posargs[::-1], example.args[::-1]))
+            example_kwargs = dict(zip(posargs[-len(example.args) :], example.args))
         else:
             example_kwargs = dict(example.kwargs)
+        given_kws = ", ".join(
+            repr(k) for k in sorted(wrapped_test.hypothesis._given_kwargs)
+        )
+        example_kws = ", ".join(repr(k) for k in sorted(example_kwargs))
+        if given_kws != example_kws:
+            raise InvalidArgument(
+                f"Inconsistent args: @given() got strategies for {given_kws}, "
+                f"but @example() got arguments for {example_kws}"
+            ) from None
 
-        bound = original_sig.bind(*arguments, **example_kwargs, **kwargs)
+        # This is certainly true because the example_kwargs exactly match the params
+        # reserved by @given(), which are then remove from the function signature.
+        assert set(example_kwargs).isdisjoint(kwargs)
+        example_kwargs.update(kwargs)
 
         if Phase.explicit not in state.settings.phases:
             continue
 
-        args, kw = convert_positional_arguments(original_sig, bound.args, bound.kwargs)
         with local_settings(state.settings):
             fragments_reported = []
             try:
                 with with_reporter(fragments_reported.append):
                     state.execute_once(
-                        ArtificialDataForExample(args, kw),
+                        ArtificialDataForExample(arguments, example_kwargs),
                         is_final=True,
                         print_example=True,
                     )

--- a/hypothesis-python/src/hypothesis/internal/reflection.py
+++ b/hypothesis-python/src/hypothesis/internal/reflection.py
@@ -186,10 +186,7 @@ def convert_positional_arguments(function, args, kwargs):
 
     new_args will only be non-empty if function has pos-only args or *args.
     """
-    if isinstance(function, inspect.Signature):
-        sig = function
-    else:
-        sig = inspect.signature(function, follow_wrapped=False)
+    sig = inspect.signature(function, follow_wrapped=False)
     bound = sig.bind(*args, **kwargs)
     new_args = []
     new_kwargs = dict(bound.arguments)

--- a/hypothesis-python/tests/cover/test_lookup_py38.py
+++ b/hypothesis-python/tests/cover/test_lookup_py38.py
@@ -9,12 +9,13 @@
 # obtain one at https://mozilla.org/MPL/2.0/.
 
 import dataclasses
+import re
 import sys
 import typing
 
 import pytest
 
-from hypothesis import given, strategies as st
+from hypothesis import example, given, strategies as st
 from hypothesis.errors import InvalidArgument, Unsatisfiable
 from hypothesis.internal.reflection import (
     convert_positional_arguments,
@@ -194,3 +195,30 @@ def test_cannot_pass_strategies_by_position_if_there_are_posonly_args():
 @given(st.booleans())
 def test_cannot_pass_strategies_for_posonly_args(x, /):
     pass
+
+
+@given(y=st.booleans())
+def has_posonly_args(x, /, y):
+    pass
+
+
+def test_example_argument_validation():
+    example(y=None)(has_posonly_args)(1)  # Basic case is OK
+
+    with pytest.raises(
+        InvalidArgument,
+        match=re.escape(
+            "Cannot pass positional arguments to @example() when decorating "
+            "a test function which has positional-only parameters."
+        ),
+    ):
+        example(None)(has_posonly_args)(1)
+
+    with pytest.raises(
+        InvalidArgument,
+        match=re.escape(
+            "Inconsistent args: @given() got strategies for 'y', "
+            "but @example() got arguments for 'x'"
+        ),
+    ):
+        example(x=None)(has_posonly_args)(1)

